src/index.ts

```diff
@@ -69,47 +77,45 @@ export class SilentPayment {
     if (silentPaymentGroups.length === 0) return ret; // passthrough
 
     const a = SilentPayment._sumPrivkeys(utxos);
-    const outpoint_hash = SilentPayment._outpointsHash(utxos);
+    const A = Buffer.from(ecc.pointFromScalar(a) as Uint8Array);
+    const outpoint_hash = SilentPayment._outpointsHash(utxos, A);
 
-    // Generating Pmn for each Bm in the group
+    // Generating Pmk for each Bm in the group
     for (const group of silentPaymentGroups) {
       // Bscan * a * outpoint_hash
       const ecdh_shared_secret_step1 = Buffer.from(ecc.privateMultiply(outpoint_hash, a) as Uint8Array);
       const ecdh_shared_secret = ecc.pointMultiply(group.Bscan, ecdh_shared_secret_step1);
 
-      let n = 0;
+      let k = 0;
       for (const [Bm, amount] of group.BmValues) {
-        const tn = crypto
-          .createHash("sha256")
-          .update(Buffer.concat([ecdh_shared_secret!, SilentPayment._ser32(n)]))
-          .digest();
+        const tk = taggedHash(
+            "BIP0352/SharedSecret",
+            Buffer.concat([ecdh_shared_secret!, SilentPayment._ser32(k)])
+        );
 
-        // Let Pmn = tn·G + Bm
-        const Pmn = Buffer.from(ecc.pointAdd(ecc.pointMultiply(G, tn) as Uint8Array, Bm) as Uint8Array);
+        // Let Pmk = tk·G + Bm
+        const Pmk = Buffer.from(ecc.pointAdd(ecc.pointMultiply(G, tk) as Uint8Array, Bm) as Uint8Array);
 
-        // Encode Pmn as a BIP341 taproot output
-        const address = Pmn.slice(1).toString("hex");
+        // Encode Pmk as a BIP341 taproot output
+        const address = Pmk.slice(1).toString("hex");
         const newTarget: Target = { address };
         newTarget.value = amount;
         ret.push(newTarget);
-        n += 1;
+        k += 1;
       }
-      n += 1;
     }
     return ret;
   }
 
-  static _outpointsHash(parameters: UTXO[]): Buffer {
+  static _outpointsHash(parameters: UTXO[], A: Buffer): Buffer {
     let bufferConcat = Buffer.alloc(0);
     const outpoints: Array<Buffer> = [];
     for (const parameter of parameters) {
       outpoints.push(Buffer.concat([Buffer.from(parameter.txid, "hex").reverse(), SilentPayment._ser32(parameter.vout).reverse()]));
     }
     outpoints.sort(Buffer.compare);
-    for (const outpoint of outpoints) {
-      bufferConcat = Buffer.concat([bufferConcat, outpoint]);
-    }
-    return crypto.createHash("sha256").update(bufferConcat).digest();
+    const smallest_outpoint = outpoints[0];
```

Comment:
what does smallest  mean here? youre just taking the first one?

Reply:
Yeah, on L116 above we sort the array of outpoints (lexicographically) and then take the first one from the sorted list (smallest one). On a previous version of the BIP we were sorting and taking all of the outpoints but after some discussion realised taking the smallest outpoint is sufficient (and makes things easier for certain use cases, like hardware wallets).

Reply:
ah yes, missed  `outpoints.sort`